BootstrapAppModelResolver returns mixed versions for a Platform test with SNAPSHOT deps

Steps to reproduce:

1. Build Quarkus revision e32ce80
2. Build Camel Quarkus revision a67ab39
3. Checkout this branch of quarkus-platform https://github.com/ppalaga/quarkus-platform/commits/200407-camel-regen
4.  Quick-build quarkus platform `mvn clean install -DskipTests`
5. Run the servlet test `cd integration-tests/camel/camel-servlet; mvn clean test`

Expected: the test should pass
Actual: the test fails with

```
java.lang.RuntimeException: java.lang.NoClassDefFoundError: io/quarkus/arc/processor/AlternativePriorities
        at io.quarkus.test.junit.QuarkusTestExtension.beforeEach(QuarkusTestExtension.java:246)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachCallbacks$1(TestMethodTestDescriptor.java:161)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeMethodsOrCallbacksUntilExceptionOccurs$5(TestMethodTestDescriptor.java:197)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeMethodsOrCallbacksUntilExceptionOccurs(TestMethodTestDescriptor.java:197)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeBeforeEachCallbacks(TestMethodTestDescriptor.java:160)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:71)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
Caused by: java.lang.NoClassDefFoundError: io/quarkus/arc/processor/AlternativePriorities
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at io.quarkus.deployment.util.ServiceUtil.classesNamedIn(ServiceUtil.java:31)
        at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:209)
        at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:136)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:102)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:240)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:128)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:52)
        at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:122)
        at io.quarkus.test.junit.QuarkusTestExtension.ensureStarted(QuarkusTestExtension.java:260)
        at io.quarkus.test.junit.QuarkusTestExtension.beforeAll(QuarkusTestExtension.java:283)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$7(ClassBasedTestDescriptor.java:359)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:359)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:189)
        at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:78)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:132)
        ... 29 more
Caused by: java.lang.ClassNotFoundException: io.quarkus.arc.processor.AlternativePriorities
        at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:322)
        at io.quarkus.bootstrap.classloading.QuarkusClassLoader.loadClass(QuarkusClassLoader.java:275)
        ... 47 more
```

Addional info:

* `io.quarkus.arc.processor.AlternativePriorities`  is available in quarkus e32ce80, but is not available in Quarkus 1.3.1.Final that Camel Quarkus a67ab39 depends on. So apparently older Quarkus deployment artifacts seem to be mixed in by BootstrapAppModelResolver.
* Indeed when outputting the tree via `BuildDependencyGraphVisitor.buildTreeConsumer`, there are a couple of 1.3.1.Final artifacts there:

```
io.quarkus:quarkus-universe-integration-tests-camel-servlet:jar:999-SNAPSHOT
├─ io.quarkus:quarkus-universe-integration-tests-camel-rpkgtests:pom:999-SNAPSHOT (compile)
├─ org.apache.camel.quarkus:camel-quarkus-integration-test-servlet:jar:1.1.0-SNAPSHOT (compile)
│  ├─ org.apache.camel.quarkus:camel-quarkus-servlet-deployment:jar:1.1.0-SNAPSHOT (compile)
│  │  ├─ io.quarkus:quarkus-undertow-deployment:jar:999-SNAPSHOT (compile)
│  │  │  ├─ io.quarkus.http:quarkus-http-servlet:jar:3.0.4.Final (compile)
│  │  │  │  └─ io.quarkus.http:quarkus-http-core:jar:3.0.4.Final (compile)
│  │  │  │     ├─ io.quarkus.http:quarkus-http-http-core:jar:3.0.4.Final (compile)
│  │  │  │     │  └─ io.netty:netty-buffer:jar:4.1.45.Final (compile)
│  │  │  │     └─ io.netty:netty-codec-http:jar:4.1.45.Final (compile)
│  │  │  │        ├─ io.netty:netty-common:jar:4.1.45.Final (compile)
│  │  │  │        ├─ io.netty:netty-transport:jar:4.1.45.Final (compile)
│  │  │  │        ├─ io.netty:netty-codec:jar:4.1.45.Final (compile)
│  │  │  │        └─ io.netty:netty-handler:jar:4.1.45.Final (compile)
│  │  │  ├─ jakarta.servlet:jakarta.servlet-api:jar:4.0.3 (compile)
│  │  │  ├─ io.quarkus:quarkus-arc-deployment:jar:999-SNAPSHOT (compile)
│  │  │  │  ├─ io.quarkus:quarkus-arc:jar:999-SNAPSHOT (compile)
│  │  │  │  │  └─ io.quarkus.arc:arc:jar:1.3.1.Final (compile)
│  │  │  │  │     └─ jakarta.transaction:jakarta.transaction-api:jar:1.3.3 (compile)
│  │  │  │  └─ io.quarkus.arc:arc-processor:jar:1.3.1.Final (compile)
│  │  │  ├─ io.quarkus:quarkus-undertow-spi:jar:1.3.1.Final (compile)
│  │  │  ├─ io.quarkus:quarkus-undertow:jar:999-SNAPSHOT (compile)
│  │  │  │  ├─ io.quarkus:quarkus-vertx-http:jar:999-SNAPSHOT (compile)
│  │  │  │  │  ├─ io.quarkus:quarkus-vertx-core:jar:999-SNAPSHOT (compile)
│  │  │  │  │  │  └─ io.quarkus:quarkus-netty:jar:999-SNAPSHOT (compile)
│  │  │  │  │  └─ io.vertx:vertx-web:jar:3.8.5 (compile)
│  │  │  │  │     ├─ io.vertx:vertx-web-common:jar:3.8.5 (compile)
│  │  │  │  │     ├─ io.vertx:vertx-auth-common:jar:3.8.5 (compile)
│  │  │  │  │     └─ io.vertx:vertx-bridge-common:jar:3.8.5 (compile)
│  │  │  │  ├─ io.quarkus.security:quarkus-security:jar:1.0.1.Final (compile)
│  │  │  │  ├─ jakarta.enterprise:jakarta.enterprise.cdi-api:jar:2.0.2 (compile)
│  │  │  │  │  ├─ jakarta.el:jakarta.el-api:jar:3.0.3 (compile)
│  │  │  │  │  └─ jakarta.interceptor:jakarta.interceptor-api:jar:1.2.5 (compile)
│  │  │  │  ├─ jakarta.annotation:jakarta.annotation-api:jar:1.3.5 (compile)
│  │  │  │  ├─ io.quarkus.http:quarkus-http-vertx-backend:jar:3.0.4.Final (compile)
│  │  │  │  │  └─ io.vertx:vertx-core:jar:3.8.5 (compile)
│  │  │  │  │     ├─ io.netty:netty-handler-proxy:jar:4.1.45.Final (compile)
│  │  │  │  │     │  └─ io.netty:netty-codec-socks:jar:4.1.45.Final (compile)
│  │  │  │  │     ├─ io.netty:netty-codec-http2:jar:4.1.45.Final (compile)
│  │  │  │  │     ├─ io.netty:netty-resolver:jar:4.1.45.Final (compile)
│  │  │  │  │     └─ io.netty:netty-resolver-dns:jar:4.1.45.Final (compile)
│  │  │  │  │        └─ io.netty:netty-codec-dns:jar:4.1.45.Final (compile)
│  │  │  │  └─ org.eclipse.microprofile.context-propagation:microprofile-context-propagation-api:jar:1.0.1 (compile)
│  │  │  ├─ io.quarkus:quarkus-vertx-http-deployment:jar:999-SNAPSHOT (compile)
│  │  │  │  └─ io.quarkus:quarkus-vertx-core-deployment:jar:999-SNAPSHOT (compile)
│  │  │  │     └─ io.quarkus:quarkus-netty-deployment:jar:999-SNAPSHOT (compile)
│  │  │  ├─ io.quarkus:quarkus-kubernetes-spi:jar:1.3.1.Final (compile)
│  │  │  └─ org.jboss.metadata:jboss-metadata-web:jar:11.0.0.Final (compile)
│  │  │     └─ org.jboss.metadata:jboss-metadata-common:jar:11.0.0.Final (compile)
│  │  ├─ org.apache.camel.quarkus:camel-quarkus-core-deployment:jar:1.1.0-SNAPSHOT (compile)
│  │  │  ├─ org.apache.camel.quarkus:camel-quarkus-support-common-deployment:jar:1.1.0-SNAPSHOT (compile)
│  │  │  │  └─ org.apache.camel.quarkus:camel-quarkus-support-common:jar:1.1.0-SNAPSHOT (compile)
│  │  │  └─ org.apache.camel.quarkus:camel-quarkus-core:jar:1.1.0-SNAPSHOT (compile)
│  │  │     ├─ org.apache.camel:camel-main:jar:3.1.0 (compile)
│  │  │     │  ├─ org.apache.camel:camel-api:jar:3.1.0 (compile)
│  │  │     │  ├─ org.apache.camel:camel-management-api:jar:3.1.0 (compile)
│  │  │     │  └─ org.apache.camel:camel-util:jar:3.1.0 (compile)
│  │  │     ├─ org.apache.camel:camel-core-engine:jar:3.1.0 (compile)
│  │  │     ├─ org.apache.camel:camel-base:jar:3.1.0 (compile)
│  │  │     ├─ org.apache.camel:camel-core-catalog:jar:3.1.0 (compile)
│  │  │     └─ org.apache.camel:camel-microprofile-config:jar:3.1.0 (compile)
│  │  │        └─ org.eclipse.microprofile.config:microprofile-config-api:jar:1.4 (compile)
│  │  └─ org.apache.camel.quarkus:camel-quarkus-servlet:jar:1.1.0-SNAPSHOT (compile)
│  │     └─ org.apache.camel:camel-servlet:jar:3.1.0 (compile)
│  │        ├─ org.apache.camel:camel-support:jar:3.1.0 (compile)
│  │        └─ org.apache.camel:camel-http-common:jar:3.1.0 (compile)
│  │           ├─ org.apache.camel:camel-http-base:jar:3.1.0 (compile)
│  │           └─ org.apache.camel:camel-attachments:jar:3.1.0 (compile)
│  │              └─ com.sun.activation:javax.activation:jar:1.2.0 (compile)
│  ├─ org.apache.camel.quarkus:camel-quarkus-rest-deployment:jar:1.1.0-SNAPSHOT (compile)
│  │  └─ org.apache.camel.quarkus:camel-quarkus-rest:jar:1.1.0-SNAPSHOT (compile)
│  │     └─ org.apache.camel:camel-rest:jar:3.1.0 (compile)
│  │        └─ org.apache.camel:camel-tooling-model:jar:3.1.0 (compile)
│  │           └─ org.apache.camel:camel-util-json:jar:3.1.0 (compile)
│  ├─ org.apache.camel.quarkus:camel-quarkus-core-cloud-deployment:jar:1.1.0-SNAPSHOT (compile)
│  │  ├─ io.quarkus:quarkus-caffeine-deployment:jar:999-SNAPSHOT (compile)
│  │  │  └─ io.quarkus:quarkus-caffeine:jar:999-SNAPSHOT (compile)
│  │  │     └─ com.github.ben-manes.caffeine:caffeine:jar:2.8.0 (compile)
│  │  │        └─ com.google.errorprone:error_prone_annotations:jar:2.2.0 (compile)
│  │  └─ org.apache.camel.quarkus:camel-quarkus-core-cloud:jar:1.1.0-SNAPSHOT (compile)
│  │     └─ org.apache.camel:camel-cloud:jar:3.1.0 (compile)
│  │        └─ org.apache.camel:camel-caffeine-lrucache:jar:3.1.0 (compile)
│  └─ org.apache.camel.quarkus:camel-quarkus-support-policy-deployment:jar:1.1.0-SNAPSHOT (compile)
│     └─ org.apache.camel.quarkus:camel-quarkus-support-policy:jar:1.1.0-SNAPSHOT (compile)
└─ org.apache.camel.quarkus:camel-quarkus-integration-test-servlet-rpkgtests:jar:1.1.0-SNAPSHOT (test)
   ├─ io.quarkus:quarkus-junit5:jar:999-SNAPSHOT (test)
   │  ├─ io.quarkus:quarkus-bootstrap-core:jar:999-SNAPSHOT (compile)
   │  │  ├─ org.ow2.asm:asm:jar:7.3.1 (compile)
   │  │  ├─ org.apache.maven:maven-embedder:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven:maven-settings:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven:maven-core:jar:3.6.3 (test)
   │  │  │  │  ├─ org.apache.maven:maven-artifact:jar:3.6.3 (test)
   │  │  │  │  └─ org.codehaus.plexus:plexus-component-annotations:jar:2.1.0 (test)
   │  │  │  ├─ org.apache.maven:maven-plugin-api:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven:maven-model:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven:maven-model-builder:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven:maven-builder-support:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven.resolver:maven-resolver-api:jar:1.4.1 (test)
   │  │  │  ├─ org.apache.maven.resolver:maven-resolver-util:jar:1.4.1 (test)
   │  │  │  ├─ org.apache.maven.shared:maven-shared-utils:jar:3.2.1 (test)
   │  │  │  │  └─ commons-io:commons-io:jar:2.6 (test)
   │  │  │  ├─ com.google.inject:guice:jar:no_aop:4.2.1 (test)
   │  │  │  │  └─ com.google.guava:guava:jar:27.0.1-jre (test)
   │  │  │  │     ├─ com.google.guava:failureaccess:jar:1.0.1 (test)
   │  │  │  │     └─ com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava (test)
   │  │  │  ├─ org.codehaus.plexus:plexus-utils:jar:3.2.1 (test)
   │  │  │  ├─ org.eclipse.sisu:org.eclipse.sisu.plexus:jar:0.3.4 (test)
   │  │  │  └─ commons-cli:commons-cli:jar:1.4 (test)
   │  │  ├─ org.apache.maven:maven-settings-builder:jar:3.6.3 (test)
   │  │  │  ├─ org.codehaus.plexus:plexus-interpolation:jar:1.25 (test)
   │  │  │  └─ org.sonatype.plexus:plexus-sec-dispatcher:jar:1.4 (test)
   │  │  │     └─ org.sonatype.plexus:plexus-cipher:jar:1.4 (test)
   │  │  ├─ org.apache.maven:maven-resolver-provider:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven:maven-repository-metadata:jar:3.6.3 (test)
   │  │  │  ├─ org.apache.maven.resolver:maven-resolver-spi:jar:1.4.1 (test)
   │  │  │  └─ org.apache.maven.resolver:maven-resolver-impl:jar:1.4.1 (test)
   │  │  ├─ org.apache.maven.resolver:maven-resolver-connector-basic:jar:1.4.1 (test)
   │  │  ├─ org.apache.maven.resolver:maven-resolver-transport-wagon:jar:1.4.1 (test)
   │  │  ├─ org.apache.maven.wagon:wagon-http:jar:3.3.4 (test)
   │  │  │  ├─ org.apache.maven.wagon:wagon-http-shared:jar:3.3.4 (test)
   │  │  │  │  └─ org.jsoup:jsoup:jar:1.12.1 (test)
   │  │  │  └─ org.apache.maven.wagon:wagon-provider-api:jar:3.3.4 (test)
   │  │  ├─ org.apache.maven.wagon:wagon-file:jar:3.3.4 (test)
   │  │  ├─ org.jboss.logging:jboss-logging:jar:3.3.2.Final (compile)
   │  │  └─ org.jboss.logging:commons-logging-jboss-logging:jar:1.0.0.Final (test)
   │  ├─ io.quarkus:quarkus-test-common:jar:999-SNAPSHOT (test)
   │  │  ├─ io.quarkus:quarkus-jsonp-deployment:jar:999-SNAPSHOT (test)
   │  │  └─ org.jboss:jandex:jar:2.1.3.Final (compile)
   │  ├─ org.junit.jupiter:junit-jupiter:jar:5.6.1 (test)
   │  │  ├─ org.junit.jupiter:junit-jupiter-api:jar:5.6.1 (test)
   │  │  │  ├─ org.apiguardian:apiguardian-api:jar:1.1.0 (test)
   │  │  │  ├─ org.opentest4j:opentest4j:jar:1.2.0 (test)
   │  │  │  └─ org.junit.platform:junit-platform-commons:jar:1.6.1 (test)
   │  │  ├─ org.junit.jupiter:junit-jupiter-params:jar:5.6.1 (test)
   │  │  └─ org.junit.jupiter:junit-jupiter-engine:jar:5.6.1 (test)
   │  │     └─ org.junit.platform:junit-platform-engine:jar:1.6.1 (test)
   │  └─ io.quarkus:quarkus-core-deployment:jar:999-SNAPSHOT (compile)
   │     ├─ org.wildfly.common:wildfly-common:jar:1.5.4.Final-format-001 (compile)
   │     ├─ io.quarkus.gizmo:gizmo:jar:1.0.2.Final (compile)
   │     │  └─ org.ow2.asm:asm-util:jar:7.3.1 (compile)
   │     │     ├─ org.ow2.asm:asm-tree:jar:7.3.1 (compile)
   │     │     └─ org.ow2.asm:asm-analysis:jar:7.3.1 (compile)
   │     ├─ io.quarkus:quarkus-development-mode-spi:jar:999-SNAPSHOT (compile)
   │     ├─ io.quarkus:quarkus-core:jar:999-SNAPSHOT (compile)
   │     │  ├─ jakarta.inject:jakarta.inject-api:jar:1.0 (compile)
   │     │  ├─ io.quarkus:quarkus-ide-launcher:jar:999-SNAPSHOT (compile)
   │     │  ├─ io.smallrye.config:smallrye-config:jar:1.7.0 (compile)
   │     │  │  └─ io.smallrye.config:smallrye-config-common:jar:1.7.0 (compile)
   │     │  ├─ org.jboss.logmanager:jboss-logmanager-embedded:jar:1.0.4 (compile)
   │     │  ├─ org.jboss.logging:jboss-logging-annotations:jar:2.1.0.Final (compile)
   │     │  ├─ org.jboss.threads:jboss-threads:jar:3.0.1.Final (compile)
   │     │  ├─ org.slf4j:slf4j-api:jar:1.7.29 (compile)
   │     │  └─ org.jboss.slf4j:slf4j-jboss-logging:jar:1.2.0.Final (compile)
   │     ├─ io.quarkus:quarkus-builder:jar:999-SNAPSHOT (compile)
   │     └─ org.graalvm.sdk:graal-sdk:jar:19.3.1 (compile)
   └─ io.rest-assured:rest-assured:jar:4.3.0 (test)
      ├─ org.codehaus.groovy:groovy:jar:3.0.2 (test)
      ├─ org.codehaus.groovy:groovy-xml:jar:3.0.2 (test)
      ├─ org.apache.httpcomponents:httpclient:jar:4.5.12 (test)
      │  ├─ org.apache.httpcomponents:httpcore:jar:4.4.13 (test)
      │  └─ commons-codec:commons-codec:jar:1.13 (test)
      ├─ org.apache.httpcomponents:httpmime:jar:4.5.3 (test)
      ├─ org.hamcrest:hamcrest:jar:2.1 (test)
      ├─ org.ccil.cowan.tagsoup:tagsoup:jar:1.2.1 (test)
      ├─ io.rest-assured:json-path:jar:4.3.0 (test)
      │  ├─ org.codehaus.groovy:groovy-json:jar:3.0.2 (test)
      │  └─ io.rest-assured:rest-assured-common:jar:4.3.0 (test)
      └─ io.rest-assured:xml-path:jar:4.3.0 (test)
         ├─ org.apache.commons:commons-lang3:jar:3.9 (test)
         ├─ jakarta.xml.bind:jakarta.xml.bind-api:jar:2.3.2 (test)
         │  └─ jakarta.activation:jakarta.activation-api:jar:1.2.1 (test)
         └─ org.apache.sling:org.apache.sling.javax.activation:jar:0.1.0 (test)
```

Strange that only those four artifacts are 1.3.1.Final and the vast majority is 999-SNAPSHOT. I'd expect all of them to be 999-SNAPSHOT.

I am not sure whether this is a misconfiguration in either of camel-quarkus, quarkus or platform or whether it is a bug in the resolver.

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/BootstrapAppModelResolver.java
@@ -192,8 +192,26 @@ public class BootstrapAppModelResolver implements AppModelResolver {
             child.accept(visitor);
         }
 
+        final ArtifactDescriptorResult appArtifactDescr = mvn.resolveDescriptor(toAetherArtifact(appArtifact));
+        if (managedDeps == null) {
+            managedDeps = appArtifactDescr.getManagedDependencies();
+        } else {
+            final List<Dependency> mergedManagedDeps = new ArrayList<>(managedDeps.size());
+            final Set<AppArtifactKey> mergedKeys = new HashSet<>(managedDeps.size());
+            for (Dependency dep : managedDeps) {
+                mergedKeys.add(getKey(dep.getArtifact()));
+                mergedManagedDeps.add(dep);
+            }
+            for (Dependency dep : appArtifactDescr.getManagedDependencies()) {
+                final Artifact artifact = dep.getArtifact();
+                if (!mergedKeys.contains(getKey(artifact))) {
+                    mergedManagedDeps.add(dep);
+                }
+            }
+            managedDeps = mergedManagedDeps;
+        }
         final List<RemoteRepository> repos = mvn.aggregateRepositories(managedRepos,
-                mvn.newResolutionRepositories(mvn.resolveDescriptor(toAetherArtifact(appArtifact)).getRepositories()));
+                mvn.newResolutionRepositories(appArtifactDescr.getRepositories()));
 
         final DeploymentInjectingDependencyVisitor deploymentInjector = new DeploymentInjectingDependencyVisitor(mvn,
                 managedDeps, repos, appBuilder);
@@ -290,6 +308,11 @@ public class BootstrapAppModelResolver implements AppModelResolver {
     public void install(AppArtifact appArtifact, Path localPath) throws AppModelResolverException {
         mvn.install(new DefaultArtifact(appArtifact.getGroupId(), appArtifact.getArtifactId(), appArtifact.getClassifier(),
                 appArtifact.getType(), appArtifact.getVersion(), Collections.emptyMap(), localPath.toFile()));
+    }
+
+    private AppArtifactKey getKey(final Artifact artifact) {
+        return new AppArtifactKey(artifact.getGroupId(), artifact.getArtifactId(),
+                artifact.getClassifier(), artifact.getExtension());
     }
 
     private String getEarliest(final VersionRangeResult rangeResult) {

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java
@@ -126,7 +126,6 @@ public class DeploymentInjectingDependencyVisitor {
         node.setData(QUARKUS_DEPLOYMENT_ARTIFACT, deploymentArtifact);
         runtimeNodes.add(node);
         Dependency dependency = new Dependency(node.getArtifact(), JavaScopes.COMPILE);
-        managedDeps.add(dependency);
         runtimeExtensionDeps.add(dependency);
         managedDeps.add(new Dependency(deploymentArtifact, JavaScopes.COMPILE));
     }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -9,6 +9,7 @@ import java.io.IOException;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
+import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -106,7 +107,7 @@ public class QuarkusTestExtension
             }
             CuratedApplication curatedApplication = runnerBuilder
                     .setTest(true)
-                    .setProjectRoot(new File("").toPath())
+                    .setProjectRoot(Files.isDirectory(appClassLocation) ? new File("").toPath() : null)
                     .build()
                     .bootstrap();
 

